### PR TITLE
Snp integration

### PR DESCRIPTION
**Pull request by smo4201**

### CODE REVIEW

File: attestationreport/attestationreport.go [hunk 1]
```diff
@@ -439,23 +436,37 @@ func Generate(nonce []byte, metadata [][]byte, measurements []Measurement, measu
 // (Signing Certificate -> Intermediate Certificates -> Root CA Certificate)
 // Parameter 'mu' is an optional mutex, in case the private key requires exclusive
 // access (e.g. because it is located in a hardware module)
-func Sign(mu *sync.Mutex, ar ArJws, priv crypto.PrivateKey, pub crypto.PublicKey, certsPem [][]byte) (bool, []byte) {
+// TODO remove func Sign(mu *sync.Mutex, ar ArJws, priv crypto.PrivateKey, pub crypto.PublicKey, certs CertChain) (bool, []byte) {
```
[the remark this answers is not in the document]
Comment (smo4201) [reply attestationreport/attestationreport.go]: Just overlooked the "TODO", removed it

**Review by jpecholt — APPROVED** [review]

File: attestationreport/attestationreport.go [hunk 2]
```diff
@@ -818,6 +827,14 @@ func VerifyJws(data string, roots *x509.CertPool, roles []string) (JwsResult, []
 				result.Summary.setFalseMulti(&msg)
 				return result, nil, false
 			}
+			if len(certs[0][0].Subject.OrganizationalUnit) == 0 {
```
[the remark this answers is not in the document]
Comment (smo4201) [reply attestationreport/attestationreport.go:830]: It was not required if "roles" were nil (line 824). If roles were not nil, hardcoded OU was the only possibility implemented so far. Anyway, I removed the roles entirely and already put the checks into the policies module which will be one of the next PRs

File: attestationreport/snp.go [hunk 3]
```diff
@@ -196,15 +202,15 @@ func decodeSnpReport(report []byte) (snpreport, error) {
 	return s, nil
 }
 
-func verifySnpParams(s *snpreport, v *Verification) (Result, bool) {
-	result := Result{
-		Success: true,
+func verifySnpVersion(s snpreport, version uint32) (Result, bool) {
+	ok := s.Version == version
```
[the remark this answers is not in the document]
Comment (smo4201) [reply attestationreport/snp.go:206]: Done

File: cmcd/main.go [hunk 4]
```diff
@@ -28,24 +28,26 @@ import (
...
-	ImaPcr        int32  `json:"imaPcr"`
-	// Key config: RSA2048 RSA4096 EC256 EC384 EC521
-	KeyConfig string `json:"keyConfig,omitempty"` // Default defined below during parsing
+	Port                  int      `json:"port"`
+	ServerAddr            string   `json:"provServerAddr"`
+	ServerPath            string   `json:"serverPath"`
+	LocalPath             string   `json:"localPath"`
+	FetchMetadata         bool     `json:"fetchMetadata"`
+	MeasurementInterfaces []string `json:"measurementInterfaces"` // TPM, SNP
+	SigningInterface      string   `json:"signingInterface"`      // TPM, TODO SNP, SW
```
[the remark this answers is not in the document]
Comment (smo4201) [reply cmcd/main.go:47]: Done

File: cmcd/main.go [hunk 5]
```diff
@@ -234,72 +176,103 @@ func main() {
...
-			TLSCert:     c.tlsCertPath,
-			DeviceSubCa: c.deviceSubCaPath,
-			Ca:          c.caPath,
-		}
-		err = tpmdriver.ProvisionTpm(c.ServerAddr+"activate-credential/", paths, certParams, c.KeyConfig)
+
+	if c.SigningInterface == "TPM" {
+		log.Info("Using TPM as Signing Interface")
+
+		// TODO put this into the gRPC Verify Call
```
[the remark this answers is not in the document]
Comment (smo4201) [reply cmcd/main.go:226]: Done

File: cmcd/main.go [hunk 6]
```diff
@@ -234,72 +176,103 @@ func main() {
...
+			log.Errorf("Failed to create new SW driver: %v", err)
 			return
 		}
-	}
 
-	certs, err := loadCerts(c)
-	if err != nil {
-		log.Errorf("Failed to load certificates: %v", err)
-		return
+		// TODO put this into the gRPC Verify Call
```
[the remark this answers is not in the document]
Comment (smo4201) [reply cmcd/main.go:254]: Done

File: cmcd/main.go [hunk 7]
```diff
@@ -234,72 +176,103 @@ func main() {
...
+	// }
 
 	serverConfig := &ServerConfig{
 		Metadata: metadata,
-		Certs:    certs,
-		Pcrs:     pcrs,
-		Roles:    roles,
-		Tpm:      &tpmdriver.Tpm{},
+		//Roles:    roles,
+		// TODO handle more then 1 CA, put this in gRPC verify call
```
[the remark this answers is not in the document]
Comment (smo4201) [reply cmcd/main.go:272]: Done

File: provserver/server.go [hunk 8]
```diff
@@ -434,6 +438,68 @@ func HandleAkCertRequest(buf *bytes.Buffer) (*bytes.Buffer, error) {
 	return &retBuf, nil
 }
 
+// HandleAkCertRequest handles an AK Cert Request (Step 2)
```
[the remark this answers is not in the document]
Comment (smo4201) [reply provserver/server.go:441]: Done

File: provserver/server.go [hunk 9]
```diff
@@ -529,6 +595,61 @@ func handleActivateCredential(writer http.ResponseWriter, req *http.Request) {
 	}
 }
 
+// HTTP handler for TPM Credential activation and Issuing of AK Certificates
+func handleSwSigning(writer http.ResponseWriter, req *http.Request) {
```
[the remark this answers is not in the document]
Comment (smo4201) [reply provserver/server.go:599]: Done

File: snpdriver/snpdriver.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+	milanUrl = "https://kdsintf.amd.com/vcek/v1/Milan/cert_chain"
+)
+
+// Snp is a structure required for implementing the Measure method
+// of the attestation report Measurer interface
+type Snp struct {
+	certChain ar.CertChain
+}
+
+func NewSnpDriver( /* TODO Config */ ) (*Snp, error) {
```
[the remark this answers is not in the document]
Comment (smo4201) [reply snpdriver/snpdriver.go:59]: Done

File: cmcd/main.go [hunk 11]
```diff
@@ -234,72 +176,89 @@ func main() {
...
 		}
+
+		measurements = append(measurements, snp)
 	}
 
-	certs, err := loadCerts(c)
-	if err != nil {
-		log.Errorf("Failed to load certificates: %v", err)
-		return
+	if c.SigningInterface == "SNP" {
```
Comment (mohuber) [inline cmcd/main.go:240]: this is not yet supported in the current PR, right? should we then remove lines 240-242 to not have an "empty if" lying around?

**Review by mohuber — APPROVED** [review]
looks great, thanks a lot!

File: cmcd/main.go (hunk 11 above)
Comment (mohuber) [reply cmcd/main.go]: done

**Pull request merged**